cmd/argo/commands/get.go

```diff
@@ -116,25 +118,301 @@ func printWorkflowHelper(wf *wfv1.Workflow) {
 		} else {
 			fmt.Fprintf(w, "%s\tPODNAME\tDURATION\tMESSAGE\n", ansiFormat("STEP", FgDefault))
 		}
-		node, ok := wf.Status.Nodes[wf.ObjectMeta.Name]
-		if ok {
-			printNodeTree(w, wf, node, 0, " ", " ")
-		}
-		onExitNode, ok := wf.Status.Nodes[wf.NodeID(wf.ObjectMeta.Name+".onExit")]
-		if ok {
-			fmt.Fprintf(w, "\t\t\t\t\n")
-			onExitNode.Name = "onExit"
-			printNodeTree(w, wf, onExitNode, 0, " ", " ")
+
+		// Convert Nodes to Render Trees
+		roots := convertToRenderTrees(wf)
+
+		// Print main and onExit Trees
+		rootNodeIDs := [2]string{wf.NodeID(wf.ObjectMeta.Name), wf.NodeID(wf.ObjectMeta.Name + "." + onExitSuffix)}
+		for _, id := range rootNodeIDs {
+			if node, ok := wf.Status.Nodes[id]; ok {
+				if root, ok := roots[node.ID]; ok {
+					root.renderNodes(w, wf, 0, " ", " ")
+				}
+			}
 		}
+
 		_ = w.Flush()
 	}
 }
 
-func printNodeTree(w *tabwriter.Writer, wf *wfv1.Workflow, node wfv1.NodeStatus, depth int, nodePrefix string, childPrefix string) {
-	nodeName := fmt.Sprintf("%s %s", jobStatusIconMap[node.Phase], node.Name)
+// Interface to represent Nodes in render form types
+type renderNode interface {
+	// Render this renderNode and its children
+	renderNodes(w *tabwriter.Writer, wf *wfv1.Workflow, depth int,
+		nodePrefix string, childPrefix string)
+	// Get the start Time for this Render Node
+	getStartTime(wf *wfv1.Workflow) metav1.Time
+	// Get the whole Node Status for this Render Node
+	getNodeStatus(wf *wfv1.Workflow) wfv1.NodeStatus
+	// Get Render node id
+	getID() string
+}
+
+// Currently this is Pod or Resource Nodes
+type executionNode struct {
+	node string
+}
+
+// Currently this is the step groups or retry nodes
+type nonBoundaryParentNode struct {
+	node     string
+	children []renderNode // Can be boundaryNode or executionNode
+}
+
+// Currently this is the virtual Template node
+type boundaryNode struct {
+	node              string
+	boundaryContained []renderNode // Can be nonBoundaryParent or executionNode or boundaryNode
+}
+
+func (nodeInfo *boundaryNode) getID() string {
```

Comment:
In general there seems to be some unnecessary repetition happening. For example, rather than redefine `getId()`, `getNodeStatus()`, `getStartTime()` for each type, you should [embed](https://golang.org/doc/effective_go.html#embedding) a common type which defines the common methods once. In this case, you would embed executionNode into nonBoundaryParentNode and boundaryNode. Then you only need to implement getID and getStartTime in executionNode and the other types will "inherit" the method.